_posts/2022-09-08-patterns-with-suspense.markdown

```diff
@@ -0,0 +1,594 @@
+layout: epic
+title: "React Suspense: What it is, how to use it, gotchas, patterns and more"
+subtitle:
+  A brief guide on how to use Suspense, and some gotchas we've encountered along
+  the way
+date: 2022-09-09
+categories: [React, Suspense]
+author: [chris]
+---
+
+With the release of React 18, React's new concurrent mode (aka
+[Suspense][suspense]) was [officially released][react]. Finally! It was first
+demoed at React Conf in [2018][reactconf], has been discussed and debated [for
+years][umbrella], and the first parts started to ship in [v16][v16]. And now,
+with React 18 shipping to a client app near you, it's time to discuss some of
+the patterns that are emerging in the community and how to avoid some common
+pitfalls that you might encounter when implementing Suspense (or refactoring)
+your UI in support of it.
+
+<!-- more -->
+
+Before we begin, it's worth noting that Suspense is a dense subject. It's more
+than just a pattern for efficient data-fetching; it's a new way to think about
+your UI by elevating concurrency to the status of first-class citizen. When
+first trying to understand how this fits together it can feel overwhelming, and
+for that reason it's important to take things slow and ideally build a simple
+app that uses the pattern. At that point you'll likely experience an "ah hah!"
+moment that renders the pattern clear and from which more complex UI
+interactions can be designed. May this post be your guide!
+
+## Definitions
+
+First, what is Suspense? Quoting the React Docs:
+
+> Suspense lets your components “wait” for something before they can render
+
+"Waiting" could mean showing a spinner while an API request takes place, a
+bundle split component is lazy-loaded via [`React.lazy`][splitting], or even a
+long-running, expensive computation is performed (without locking up the UI).
+It's a way to declaratively state that in some fashion or another _we need to
+explicitly manage time_.
+
+To demonstrate, imagine an app that renders logged-in user info along with a
+list of artists and artworks that they follow:
+
+```tsx
+const App = () => {
+  return (
+    <Suspense fallback="Loading...">
+      <UserInfo />
+
+      <Suspense fallback="Fetching Artists...">
+        <FollowedArtists />
+      </Suspense>
+      <Suspense fallback="Fetching Artworks...">
+        <FollowedArtworks />
+      </Suspense>
+    </Suspense>
+  )
+}
+```
+
+A couple things are happening here that are important to pay attention to, most
+notably the structure of the component tree and the nesting of various Suspense
+boundaries. Breaking it down a bit:
+
+1. `<App />` is wrapped in a top-level `<Suspense />` boundary. This tells React
+   that it is waiting for something to happen
+1. The first component nested within the top-level suspense boundary is
+   `<UserInfo />`. React attempts to render it
+1. Inside of `<UserInfo />` there's a mechanism for fetching user data. When
+   triggered, this tells React that we need to "suspend"
```

Comment:
I found this is a bit confusing since we haven't seen `<UserInfo />` at this point - maybe we could just add something like this to clarify: 

```suggestion
1. Inside of `<UserInfo />` there's a mechanism for fetching user data (you'll see how
   that works below. When triggered, this tells React that we need to "suspend"
```